lms/djangoapps/program_enrollments/api/v1/views.py

```diff
@@ -345,150 +345,162 @@ def post(self, request, *args, **kwargs):
+            external_user_key=student_key,
+            program_uuid=program_uuid,
+            curriculum_uuid=request_data.get('curriculum_uuid'),
+            status=request_data.get('status')
+        )
+        return enrollment.status
+
+    # pylint: disable=unused-argument
+    def modify_program_enrollment(self, request_data, program_uuid, program_enrollment):
         """
-        Helper method that returns an OrderedDict of rows from request.data,
-        keyed by the `external_user_key`.
+        Change the status of an existing program enrollment
         """
-        return OrderedDict((
-            row.get(REQUEST_STUDENT_KEY),
-            {
-                'program_uuid': program_uuid,
-                'curriculum_uuid': row.get('curriculum_uuid'),
-                'status': row.get('status'),
-                'external_user_key': row.get(REQUEST_STUDENT_KEY),
-            })
-            for row in request.data
-        )
+        if not program_enrollment:
+            return ProgramEnrollmentResponseStatuses.NOT_IN_PROGRAM
+
+        program_enrollment.status = request_data.get('status')
+        program_enrollment.save()
+        return program_enrollment.status
 
-    def _remove_existing_entries(self, program_uuid, student_data):
-        """ Helper method to remove entries that have existing ProgramEnrollment records. """
-        result = {}
-        existing_enrollments = ProgramEnrollment.bulk_read_by_student_key(program_uuid, student_data)
-        for enrollment in existing_enrollments:
-            result[enrollment.external_user_key] = CourseEnrollmentResponseStatuses.CONFLICT
-            student_data.pop(enrollment.external_user_key)
-        return result
-
-    def _get_created_or_updated_response(
-            self, request, created_or_updated_data, response_data, default_status=status.HTTP_201_CREATED
-    ):
+    def create_or_modify_program_enrollment(self, request_data, program_uuid, program_enrollment):
+        if program_enrollment:
+            return self.modify_program_enrollment(request_data, program_uuid, program_enrollment)
+        else:
+            return self.create_program_enrollment(request_data, program_uuid, program_enrollment)
+
+    def get_existing_program_enrollments(self, program_uuid, student_data):
+        """ Returns the existing program enrollments for the given students and program """
+        student_keys = [data['student_key'] for data in student_data]
+        return {
+            e.external_user_key: e
+            for e in ProgramEnrollment.bulk_read_by_student_key(program_uuid, student_keys)
+        }
+
+    def _get_created_or_updated_response(self, response_data, default_status=status.HTTP_201_CREATED):
         """
         Helper method to determine an appropirate HTTP response status code.
         """
         response_status = default_status
-
-        if not created_or_updated_data:
+        good_count = len([
+            v for _, v in response_data.items()
```

Comment:
Nit: you can iterate just the values of a dict: `v for v in response_data.values() if ...`